### PR TITLE
Add eternascript submit_solution API

### PR DESCRIPTION
**Pull request by luxaritas**
## Summary
This provides script authors (eg, for the mutation/submission booster) a convenient and reliable way to submit solutions (without manually constructing a POST request, which is awkward and could result in incorrect validation or missing information (eg, the undoblock with the custom target not being included)

## Implementation Notes
* For multiple reasons, this has to be async. Promises seemed like the cleanest and most flexible way to do this.
* This API distinguishes between "hard errors" (ie, required validation failed, the backend returned an error, etc) which actually reject the promise and "user cancellation" which causes the promise to resolve to `false`. This allows for script authors to handle these situations differently ("this errored" vs "the user was prompted and opted to bail" are semantically different)
* There are two options that can be specified to adjust the behavior. The first is `validate`. If false, the validation steps that are user-bypassable (eg, "this constraint hasnt been satisfied") will be skipped. The other is "notifyOnError". If false, the promise will be immediately resolved with false instead of prompting if validate is true and validation fails, and also avoids surfacing informational dialogs that would normally occur if an error is returned from the backend (eg, for an in-use barcode). I have defaulted both of these to true, as that makes it less likely to result in undesired behavior if the options are not fully understood by the author (and is less likely to create confusion for the end-user).
* submitCurrentPose, which is called by both submit_solution and as a callback for UI submission handling, is set to use the parameters required by the UI callbacks by default

## Testing
Tested a variety of combinations of successful and unsuccessful submissions from the UI and via the Eternascript function (with various combinations of parameters)

## Related Issues
Related discussion in https://forum.eternagame.org/t/is-there-a-working-bulk-submission-script/4552/

### CODE REVIEW

**Review by tkaragianes — APPROVED** [review]
Seems all right, although the exploding conditionals to deal with error state isn't the easiest to track. Most of my concerns relate to adding comments to explain the intent behind the pipeline stages. As long as we document the submission API, I think it's fine from a script author DX perspective.

File: src/eterna/mode/PoseEdit/PoseEditMode.ts [hunk 1]
```diff
@@ -2449,15 +2499,20 @@ export default class PoseEditMode extends GameMode {
...
-                const dialog = this.showNotification(data['error'].replace(/ +/, ' '), 'More Information');
-                dialog.extraButton.clicked.connect(() => window.open(EternaURL.BARCODE_HELP, '_blank'));
-                const hairpin = this._puzzle.getBarcodeHairpin(seq);
-                SolutionManager.instance.addHairpins([hairpin.sequenceString()]);
-                this.checkConstraints();
-            } else {
-                this.showNotification(data['error']);
+            if (errorHandling.notify) {
+                let dialog;
+                if (data['error'].indexOf('barcode') >= 0) {
```
[the remark this answers is not in the document]
Comment (luxaritas) [reply src/eterna/mode/PoseEdit/PoseEditMode.ts:2504]: There is one particular error returned by the backend that indicates a barcode is already in use. This detects that error and adds some extra info to the error dialog and makes an internal note that the barcode is in use for when it checks currently used hairpins for the constraint

[the remark this answers is not in the document]
Comment (luxaritas) [reply src/eterna/mode/PoseEdit/PoseEditMode.ts:2504]: Ooh that said, if that error shows up, it should _always_ track the used barcode, not just when notifications are enabled

**Comment by tkaragianes** [comment]
On further reflection, if the main use case is bulk submission of a large set of sequences, should the defaults be user intervention required? Like `details = prompt` and `notifyOnError: true`. I guess the argument would be that by default, we want to surface issues to the user, and put the onus on the script author to appropriately call the API in a bulk submission situation.

**Comment by luxaritas** [comment]
Yes, exactly - it's trivial for the author to use the correct parameters, but incorrectly-used defaults could lead to unintuitive behavior. In the case of `details`, they most likely would need to specify it _anyways _ (this default just makes it work more reasonably if they don't), and I could forsee it being entirely possible that in the bulk submission case it may be desirable for either `notifyOnError` or `validate` to be enabled with the other disabled (but enabling both is "safe", and I'm not going to pretend to know which permutation would be desirable anyways)

File: src/eterna/mode/PoseEdit/PoseEditMode.ts [hunk 2]
```diff
@@ -1436,6 +1437,18 @@ export default class PoseEditMode extends GameMode {
                 // }
                 // return res;
             });
+
+            this._scriptInterface.addCallback('submit_solution', async (
+                details: {title?: string, description?: string} | 'prompt' = 'prompt',
+                options?: {
+                    validate?: boolean,
```
[the remark this answers is not in the document]
Comment (luxaritas) [reply src/eterna/mode/PoseEdit/PoseEditMode.ts:1444]: Probably not here, but a docstring for submitCurrentPose is a good idea

**Pull request merged**